pkg/backup/scheduler.go

```diff
@@ -0,0 +1,112 @@
+package backup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	logger "github.com/rs/zerolog/log"
+)
+
+var log = logger.With().Str("component", "backup").Logger()
+
+// Scheduler executes backups at a regular interval.
+type Scheduler struct {
+	Frequency      int // in minutes
+	NotificationCh chan bool
+
+	backuper       *Backuper
+	notify         bool
+	tickerInterval time.Duration
+
+	// control
+	close     chan struct{}
+	closeOnce sync.Once
+}
+
+// BackuperOptions options needed to instantiate a backuper.
+type BackuperOptions struct {
+	SourcePath, BackupDir string
+	Opts                  []Option
+}
+
+// NewScheduler creates a new backup scheduler.
+func NewScheduler(frequency int, opts BackuperOptions, notify bool) (*Scheduler, error) {
+	if frequency < 1 || frequency >= 1440 {
+		return nil, errors.New("interval should be in [1,1440)")
+	}
+
+	backuper, err := NewBackuper(opts.SourcePath, opts.BackupDir, opts.Opts...)
+	if err != nil {
+		return nil, fmt.Errorf("new backuper: %s", err)
+	}
+
+	return &Scheduler{
+		Frequency:      frequency,
+		NotificationCh: make(chan bool),
+
+		notify:   notify,
+		backuper: backuper,
+		close:    make(chan struct{}),
+
+		tickerInterval: time.Minute,
+	}, nil
+}
+
+// Run starts the scheduler and listens for a shutdown call.
+func (s *Scheduler) Run() {
+	log.Info().Msg("starting backup scheduler")
+
+	// wait until next interval to start
+	now, interval := time.Now(), time.Duration(s.Frequency)*s.tickerInterval
```

Comment:
I did a very simple implementation of scheduling. It figures out that the next time the backup should be run according to the configurable frequency. It will wait until then to execute the first backup. 

After that, it executes the backup and waits until the next time again.

Reply:
`wait := now.Round(interval).Sub(now)` may be cleaner